### ISSUE TITLE
Follow the "new algorithm check-list" for EC J-PAKE

### ISSUE THREAD

**Issue by mpg** [Mbed-TLS/mbedtls#6504]
An implementation of EC J-PAKE in PSA has been merged without any record of the check-list in #5904 being checked.

This task is to follow the check-list, and if any steps were missed, either execute them or make sure they're tracked (new issue, or expanded definition of an existing issue), otherwise state for the record that we've checked.

**Comment by AndrzejKurek** [comment]
I have done an analysis of:
Ad-hoc KDF:               https://github.com/Mbed-TLS/mbedtls/pull/6115
Expose EC J-PAKE:         https://github.com/Mbed-TLS/mbedtls/pull/5871
Fix EC J-PAKE format:     https://github.com/Mbed-TLS/mbedtls/pull/6390
Use EC J-PAKE in TLS 1.2: https://github.com/Mbed-TLS/mbedtls/pull/6533
Based on https://github.com/Mbed-TLS/mbedtls/pull/5904 and categorized it in similar topics.
**Questions/actions in bold.**

- PSA Crypto API draft - EC J-PAKE pre-existing (did not check if implementation is compliant), Ad-hoc KDF addition in progress here: https://github.com/ARM-software/psa-crypto-api/issues/564
- New macros and functions in `crypto_values.h` or `crypto_extra.h`:
    - PSA_ALG_IS_TLS12_ECJPAKE_TO_PMS was removed as there's no transformation to be done;
	- PSA_ALG_TLS12_ECJPAKE_TO_PMS added to crypto_values.h;
	- PSA_PAKE_OUTPUT_SIZE, PSA_PAKE_INPUT_SIZE, PSA_PAKE_OUTPUT_MAX_SIZE, PSA_PAKE_INPUT_MAX_SIZE, PSA_PAKE_OPERATION_INIT, MBEDTLS_PSA_PAKE_BUFFER_SIZE defined in crypto_extra.h;
	
**The checklist mentions that new constants must have a test case in test_suite_psa_crypto_metadata.data to verify that PSA_ALG_IS_XXX macros behave properly, but we decided to not create one. Should we add it to be compliant?
We should probably add `PSA_ALG_TLS12_ECJPAKE_TO_PMS` metadata tests to test_suite_psa_crypto_metadata.data, as mentioned in https://github.com/Mbed-TLS/mbedtls/blob/18e76e7eefd484c5f14d64beaeee98d70b12cee9/docs/architecture/psa-crypto-implementation-structure.md#new-functions-and-macros**

- Preprocessor symbols: 
    - PSA_WANT_ALG_TLS12_ECJPAKE_TO_PMS, PSA_WANT_ALG_JPAKE added to crypto_config.h;
	- No driver dispatch handling added;
	- **Should there be a check added in check_crypto_config.h that would ensure that if PSA_WANT_ALG_JPAKE is defined PSA_WANT_ALG_TLS12_ECJPAKE_TO_PMS also has to be defined?**
	
- config_psa.h - PSA_WANT_ALG_JPAKE requirements added on one side, on other MBEDTLS_ECJPAKE_C influence on MBEDTLS_PSA_BUILTIN_PAKE, MBEDTLS_PSA_BUILTIN_ALG_JPAKE and PSA_WANT_ALG_JPAKE too.
**Should enabling `MBEDTLS_ECJPAKE_C` also make `MBEDTLS_PSA_BUILTIN_ALG_TLS12_ECJPAKE_TO_PMS` defined?**

- Implementations added to psa_crypto_pake.c (new file), dispatch added to psa_crypto.c.
**Should `psa_key_derivation_tls12_ecjpake_to_pms_read`, `psa_tls12_ecjpake_to_pms_input` be moved to `psa_crypto_pake.c`?**

- crypto_extra.h contains `psa_pake_operation_s`, which contains some implementation-specific fields. **Should this be in `include/psa/crypto_builtin_*` instead?**

- Tests added to test_suite_psa_crypto, review should determine if the volume is sufficient. 

- crypto_knowledge.py - JPAKE added as its own category, TLS12_ECJPAKE_TO_PMS added - **we could consider adding an abbreviation to "short_expression", as even `_PSK_TO_MS` is abbreviated to `_PSK2MS`, and we're introducing `TLS12_ECJPAKE_TO_PMS`, which is longer.**

- Changelog: TLS12_ECJPAKE_TO_PMS has it's own entry, same for exposing the EC J-PAKE functionality through the PSA API. **We should add an entry once https://github.com/Mbed-TLS/mbedtls/pull/6533 is merged to signal that it's usable. Maybe it should be a part of this PR?**

Please note that I didn't check if everything is compliant with the PSA EC J-PAKE spec, or if the functions are compliant with PSA guidelines in general - I assumed that this was done when reviewing respective PR's.

**Comment by gilles-peskine-arm** [comment]
> We should probably add PSA_ALG_TLS12_ECJPAKE_TO_PMS metadata tests to test_suite_psa_crypto_metadata.data,

Yes, and `PSA_ALG_JPAKE`. And `ALG_IS_PAKE` to the `.function` file.

> Should there be a check added in check_crypto_config.h that would ensure that if PSA_WANT_ALG_JPAKE is defined PSA_WANT_ALG_TLS12_ECJPAKE_TO_PMS also has to be defined?

I don't see why. Those are independent algorithms. But there should be a check that requires `PSA_WANT_ALG_TLS12_ECJPAKE_TO_PMS` as well as `PSA_WANT_ALG_JPAKE` when TLS has support for ECJPAKE via PSA.

> Should enabling MBEDTLS_ECJPAKE_C also make MBEDTLS_PSA_BUILTIN_ALG_TLS12_ECJPAKE_TO_PMS defined?

Again, no: they're independent. Also, builtin algorithms shouldn't be turned on automatically: they should only be pulled in if `PSA_WANT && !MBEDTLS_PSA_ACCEL`.

> Should psa_key_derivation_tls12_ecjpake_to_pms_read, psa_tls12_ecjpake_to_pms_input be moved to psa_crypto_pake.c?

Not really important, but I think the key derivation functions should stay with the other key derivation functions (`psa_crypto.c`, with the implementation to be moved to a new file when we add driver support).

> Should this be in include/psa/crypto_builtin_* instead?

Yes, this will be done as part of the work on driver support.

> we could consider adding an abbreviation to "short_expression", as even _PSK_TO_MS is abbreviated to _PSK2MS, and we're introducing TLS12_ECJPAKE_TO_PMS, which is longer.

Possibly. In an ideal world there wouldn't be such abbreviations, but there's a limit on the length of test case descriptions and these are automatically constructed from pieces that add up.

> We should add an entry once Use PSA EC-JPAKE in TLS (1.2) - Part 2 #6533 is merged to signal that it's usable. Maybe it should be a part of this PR?

Yes. No opinion on which PR.

**Comment by AndrzejKurek** [comment]
> Yes, and `PSA_ALG_JPAKE`. And `ALG_IS_PAKE` to the `.function` file.

There are already tests for it [here](https://github.com/Mbed-TLS/mbedtls/blob/4cf77e99ab43105b863061c85796b8dbffd93ab1/tests/suites/test_suite_psa_crypto_metadata.function#L612) and [here](https://github.com/Mbed-TLS/mbedtls/blob/4cf77e99ab43105b863061c85796b8dbffd93ab1/tests/suites/test_suite_psa_crypto_metadata.data#L329).

Testing improvements added here: https://github.com/Mbed-TLS/mbedtls/pull/6652
Changelog to be added separately for the entirety of EC J-PAKE changes.

**Comment by AndrzejKurek** [comment]
Changelog issue added here: https://github.com/Mbed-TLS/mbedtls/issues/6653
If you're satisfied with the outcome and think that there's nothing more to discuss @mpg, @gilles-peskine-arm, I believe that this issue can be closed.
